src/connectors/impls/gbq/writer/sink.rs

```diff
@@ -368,93 +367,114 @@ where
+                serialized_rows = Vec::with_capacity(event.len() - row_count);
+                size = 0;
+            }
+
+            size += serialized_event.len();
         }
 
-        let request = AppendRowsRequest {
-            write_stream: write_stream.name.clone(),
-            offset: None,
-            trace_id: "".to_string(),
-            rows: Some(append_rows_request::Rows::ProtoRows(ProtoData {
-                writer_schema: Some(ProtoSchema {
-                    proto_descriptor: Some(mapping.descriptor().clone()),
-                }),
-                rows: Some(ProtoRows { serialized_rows }),
-            })),
-        };
-        let req_timeout = Duration::from_nanos(self.config.request_timeout);
-        let append_response = client
-            .append_rows(stream::iter(vec![request]))
-            .timeout(req_timeout)
-            .await;
+        request_data.push(serialized_rows);
 
-        let append_response = if let Ok(append_response) = append_response {
-            append_response
-        } else {
-            // timeout sending append rows request
-            error!(
-                "{ctx} GBQ request timed out after {}ms",
-                req_timeout.as_millis()
-            );
-            ctx.notifier.connection_lost().await?;
+        if request_data.len() > 1 {
+            warn!("The batch is too large to be sent in a single request, splitting it into {} requests. Consider lowering the batch size.", request_data.len());
+        }
 
-            return Ok(SinkReply::FAIL);
-        };
-        if let Ok(x) = append_response?
-            .into_inner()
-            .next()
-            .timeout(req_timeout)
-            .await
-        {
-            match x {
-                Some(Ok(res)) => {
-                    if let Some(updated_schema) = res.updated_schema.as_ref() {
-                        let fields = updated_schema
-                            .fields
-                            .iter()
-                            .map(|f| {
-                                format!(
-                                    "{}: {:?}",
-                                    f.name,
-                                    TableType::from_i32(f.r#type).unwrap_or_default()
-                                )
-                            })
-                            .collect::<Vec<_>>()
-                            .join("\n");
-
-                        info!("{ctx} GBQ Schema was updated: {}", fields);
-                    }
-                    if let Some(res) = res.response {
-                        match res {
-                            Response::AppendResult(AppendResult { .. }) => Ok(SinkReply::ACK),
-                            Response::Error(e) => {
-                                error!("{ctx} GBQ Error: {} {}", e.code, e.message);
-                                Ok(SinkReply::FAIL)
+        for serialized_rows in request_data {
+            let request = AppendRowsRequest {
```

Comment:
just a stupid idea to get correct estimates: build an AppendRowsRequest at the start and obtain a `&mut` reference to the `ProtoRows::serialized_rows` vec that we push to for every serialized event. Then we can check with https://docs.rs/prost/latest/prost/trait.Message.html#tymethod.encoded_len the correct size of the request. But i guess this will be a hell of a mess with the borrow checker.